### ISSUE TITLE
Disambiguating uncaught errors

### ISSUE THREAD

**Issue by jugglinmike** [w3c/web-platform-tests#10853]
Single page tests are convenient because they shorten test code by two lines (i.e. `test(function(){` and `});`). However, they also introduce an ambiguity which makes differentiating harness errors and harness errors difficult (or maybe impossible).

The ambiguity is apparent any time a test produces an uncaught error without defining any tests. This may represent one of two states:

- a valid "single page test" that has failed
- a test that has been invalidated by exceptional behavior outside of a sub-test

At this moment, it is not possible for testharness.js to differentiate between these two states. Today, it conservatively assumes the former. This is a problem because it means WPT cannot detect (or reject) tests that are invalid in this way. gh-10706 is the most recent example we have identified.

As far as I know, "single page tests" are only offered as a convenience to test authors. The ambiguity described above trumps the convenience, so if the motivation really is so simple, then we should remove the feature.

There may be cases where the semantics under test necessitate that functionality. In that case, we could preserve the feature and eliminate the ambiguity by requiring an explicit "opt-in" pattern. This might take the form of metadata in the test code (e.g. `// META: single=true`) or a filename-based feature pattern (e.g. `.single.html`).

In either case, all "single page tests" defined in WPT today would need to be updated. [My attempts to estimate the number of single-page tests have been woefully misguided.](https://w3.logbot.info/testing/20180504#c403285-c403319) I think I'll need to instrument WPT and do a complete run to get an accurate number. Before investing that time, I'd like to get feedback from the folks who would know better than me.

@jgraham @gsnedders @zcorpan @foolip What do you think about eliminating single page tests? Or requiring explicit opt-in?

**Comment by jugglinmike** [comment]
Would you consider requiring metadata as undermining the motivation for the feature?

**Comment by jgraham** [comment]
Yes, I would expect people to forget, it is by definition impossible to lint for (or we could just use that criteria rather than the metadata), and in any case requiring extra verbosity in the mode specifically designed to reduce verbosity seems to defeat the point.

**Comment by jugglinmike** [comment]
> I would expect people to forget, it is by definition impossible to lint for (or we could just use that criteria rather than the metadata)

As part of this change, `testharness.js` could report a harness error for any test that did not define at least one sub-test. If a contributor expecting single-page test behavior forgot to include metadata, their test would not receive an "OKAY" status in any browser.

As a side-benefit, this would also improve results accuracy for tests that only define sub-tests if the implementation is available. [Currently, tests like that simply time out.](https://wpt.fyi/web-animations/animation-model/animation-types/interpolation-per-property.html?sha=6736e3f46b)

This discussion has made me realize that this can't be implemented with the `// META` pattern I initially suggested. It still seems possible to do it via file name, but I don't like cramming so much information into the file name. Another alternative for opting in would be via a new property passed to the global `setup` function.

> in any case requiring extra verbosity in the mode specifically designed to reduce verbosity seems to defeat the point

If we were trying to save two lines of code as I originally suggested, then I would agree. Your feedback that the savings can be more substantial (both in line count and in cognitive overhead) is well-taken, though. I don't think the additional configuration detracts from this.

**Comment by zcorpan** [comment]
We could have an easier to remember opt in like `single()`

**Comment by zcorpan** [comment]
I made an issue for that, https://github.com/w3c/web-platform-tests/issues/10877

**Comment by jgraham** [comment]
I really dislike the idea of forcing an opt-in syntax on people who are intentionally using the features designed to reduce verbosity. From a global perspective I think it makes sense to favour the needs of test authors over the needs of people working on infrastructure.

I can imagine putting information about whether the test used single-page mode in the results, so sufficiently clever infrastructure could infer that something producing a single subtest with an ERROR in a subset of browsers, and in single page mode, but multiple or different subtests in other browsers and not in single-page-mode is probably actually an uncaught exception. Or it might be possible to propogate the error to the harness for a single page test where no `assert` function was called, but I didn't think too hard about that.

**Comment by jugglinmike** [comment]
> I really dislike the idea of forcing an opt-in syntax on people who are intentionally using the features designed to reduce verbosity.  From a global perspective I think it makes sense to favour the needs of test authors over the needs of people working on infrastructure.

As noted above, the implicit behavior also allows people to use the feature *unintentionally*. In other words: this would also serve the needs of test authors.

That said, it seems like an over-simplification to label this issue a conflict between test authors and infrastructure maintainers. That suggests that the issue is a subjective matter of convenience between the two groups. The issue we're discussing is more fundamental because it concerns ambiguity in the way tests are interpreted. It limits the project's ability to identify errors, and that effects test authors just as much as infrastructure maintainers.

**Comment by zcorpan** [comment]
I'll also note that an opt-in makes the feature discoverable and searchable, which serves test authors yet unfamiliar with the feature.

**Comment by jgraham** [comment]
I think that something like #11024 achieves what you want here.

**Comment by foolip** [comment]
As @zcorpan said on the PR, it's an improvement, but the ability to actually have tests fail when there are exceptions that *should* fail the test would be even better. I think a one-liner opt-in like `single()`, `file_is_test()` or `<meta something>` would be better overall. Perhaps this is because I never liked single-page tests and was always confused when I saw them in review, but how many people really understand that we have this feature in testharness.js and how to use it correctly?

**Comment by jgraham** [comment]
I don't think that adding extra syntax really helps people understand the feature or helps them use it correctly. On the other hand it does directly undermine the major reason for having the feature which is to reduce boilerplate in cases where that is possible.

I am much more inclined to proceed with the zero-boilerplate improvements to address the most serious issues than to take this approach.

**Comment by domenic** [comment]
I was asked for my opinion as a test author in IRC.

I am happy to use a small opt-in of this sort if I use single-page tests. However, I have been leaning away from single-page tests because it's already hard enough to remember to add `done()` for sync tests or set the `<title>` appropriately to get good error messages.

**Comment by zcorpan** [comment]
How do you prevent reviewers from being confused when reviewing single page tests, when the reviewer is not familiar with the feature? How do you prevent tests from accidentally invoking the feature (a failing assert outside test) and fail to report the intended tests?

I understand that the use case is to minimize boilerplate, but I think this is pretty minor overhead and there are several advantages.

**Comment by jgraham** [comment]
I think that the scope of #10877 is necessarily smaller; I don't think we can block tests on `ERROR`, just have a warning in that case. Even if we could it would only catch cases that happen to error in the browser configurations we run in CI.

In general I think the desire to get rid of all `ERROR`s is over-focussing on getting clean metrics and won't provide value proportional to the effort required. A test `ERROR` isn't a sure sign of a bug in the test, and there are plenty of tests that *are* buggy but have some non-ERROR status. As a result browser developers have to look at `ERROR` and `FAIL` equally. The situation is of course very different for tests that `ERROR` or `TIMEOUT` across multiple browsers, including ones that are expected to implement a feature — which is the scope of #11269 — but again, the same is true of `FAIL`; it's just there are fewer tests that `ERROR` so it seems like a more tractable problem to start by cleaning up the legacy of bad tests there.

It *is* annoying if a test throws an error that prevents it running some subtests that would otherwise have passed. But since that case explicitly doesn't apply to single-page tests, for the purposes of this issue it just balances the fact that it's slightly easier to end up with `ERROR` in a single page test than in a normal test.

It's important to remember that web-platform-tests still have relatively limited visibility for browser developers; for example we *still* don't have a dashboard showing test results in development versions. Given that, it's unsurprising that tests written 10 years ago for features that aren't in active devlopment haven't recieved much attention and are often rather broken. As the tests become more tightly integrated into development workflows and the visibility of results in multiple browsers continues to improve, I would expect the overall quality of the tests to trend upwards. I expect there to be a lot fewer tests written in 2018 that are broken in 2025 than there are tests written in 2011 broken today.

I think the idea of falling back to the filename if `<title>` is missing is good, and would be happy to accept a PR for that.

**Comment by zcorpan** [comment]
Title is https://github.com/web-platform-tests/wpt/pull/11403 -- I'll also do another PR to get the title from `// META: title=` in the worker case.

**Comment by foolip** [comment]
This was a problem in https://github.com/web-platform-tests/wpt/pull/15577, where an error in setup script caused a test to be treated as a single page test, clearly unintentionally. With an opt in to single page tests, that would instead have been treated as a harness error.

Deleting these lines would also solve the problem, however:
https://github.com/web-platform-tests/wpt/blob/a9f087c1dc0bec1f322339926f7056d0ac47cd0b/resources/testharness.js#L3304-L3306
